processor/src/main/java/org/jboss/pnc/processor/ClientGenerator.java

```diff
@@ -204,10 +143,107 @@ private boolean process0(Set<? extends TypeElement> annotations, RoundEnvironmen
         return true;
     }
 
-    private void addDefaultParameters(ExecutableElement restApiMethod, MethodSpec.Builder builder) {
+    private MethodSpec buildMethod(ExecutableElement restApiMethod, boolean withParameters) {
+        String parametersList = getParameters(restApiMethod);
+
+        TypeMirror returnType = restApiMethod.getReturnType();
+        TypeMirror returnGeneric = null; //TODO check for null before adding a generic (currently no such case)
+        if (returnType instanceof DeclaredType) {
+            if (((DeclaredType)returnType).getTypeArguments().size() > 0) {
+                returnGeneric = ((DeclaredType)returnType).getTypeArguments().get(0);
+            }
+        }
+
+        MethodSpec.Builder builder = MethodSpec.methodBuilder(restApiMethod.getSimpleName().toString())
+                .addModifiers(Modifier.PUBLIC);
+
+        builder.addException(ClassName.get("org.jboss.pnc.client", "RemoteResourceException"))
+                .beginControlFlow("try");
+
+        if (withParameters) {
+            builder.addStatement("PncClientRequestParams.setParams(parameters)");
+        }
+
+        //startsWith because off the generics
+        if (ClassName.get(returnType).toString().startsWith(ClassName.get("org.jboss.pnc.dto.response", "Page").toString())) {
+            //Page result
+            List<String> parameters = new ArrayList<>();
+            for (VariableElement parameter : restApiMethod.getParameters()) {
+                if (!ClassName.get(parameter.asType()).toString().equals("org.jboss.pnc.rest.api.parameters.PageParameters")) {
+                    builder.addParameter(TypeName.get(parameter.asType()), parameter.getSimpleName().toString());
+                    parameters.add(parameter.getSimpleName().toString());
+                } else {
+                    parameters.add("pageParameters");
+                }
+            }
+            if (withParameters) {
+                builder.addParameter(MAP_OF_STRINGS, "parameters");
+            }
+            String endpointInvokeParameters = String.join(", ", parameters);
+
+            ClassName className = ClassName.get("org.jboss.pnc.client", "RemoteCollection");
+            builder.returns(ParameterizedTypeName.get(className, TypeName.get(returnGeneric)));
+            builder
+            .addStatement("PageReader pageLoader = new PageReader<>((pageParameters) -> getEndpoint()." + restApiMethod.getSimpleName() + "(" + endpointInvokeParameters + "), getRemoteCollectionConfig())")
+            .addStatement("return pageLoader.getCollection()");
+        } else if (ClassName.get(returnType).toString().startsWith(ClassName.get("org.jboss.pnc.dto.response", "Singleton").toString())) {
+            //single result
+            addDefaultParameters(restApiMethod, builder, withParameters);
```

Comment:
This line is not nice to read: "addDefaultParameters" "withParametrs". Maybe rename variable to reflect what kind of extra parameters are added.

Reply:
I'm changing to `withQueryParams`